docs/RFCS/20181209_lazy_txn_record_creation.md

```diff
@@ -0,0 +1,206 @@
+- `txnwait.Queue` could have an explicit code path to deal with a missing transaction record for the pushee that takes the provided `TxnMeta` object into account without explicitly synthesizing a record.
+
+This current thinking is that the first alternative should be pursued.
+
+### Replay Protection / Idempotency
+
+Transaction replays are a serious concern. If a batch or a series of batches gets duplicated or lost and later shows up, we can't allow it to have adverse effects. Currently, this is guarded against by updating the write timestamp cache on `EndTransaction` requests and consulting the write timestamp cache on `BeginTransaction` requests. Because the old transaction state machine required a transaction record to commit and that a `BeginTransaction` request would fail if it ran into the write timestamp cache, this prevented replays of transactions.
+
+Now that we no longer have `BeginTransaction` requests, things are a little different. However, the same approach can be used. We can update the write timestamp cache on `EndTransaction` requests and consult it on committing `EndTransaction` requests. This will prevent more than a single `EndTransaction` request from executing successfully.
+
+However, we should be able to do more. We can use MVCC to enforce this replay protection for us. If the EndTransaction request is replayed alone then it is harmless to let it write a new `COMMITTED` transaction record. If any writes are replayed after a successful commit then they will necessarily hit WriteTooOld errors and will be pushed to a new timestamp. If those writes are in the same batch as the EndTransaction then they will prevent it from committing. If those writes are in different batches from the EndTransaction then they will not be resolvable.
+
+| Initial Action               | Replay                                | Protection                                                   |
+|------------------------------|---------------------------------------|--------------------------------------------------------------|
+| txn aborted by coordinator   | replay commit                         | not allowed by client, see [TxnCoordSender](#TxnCoordSender) |
+| txn aborted by coordinator   | replay abort                          | not an issue                                                 |
+| txn committed by coordinator | replay abort                          | not allowed by client, see [TxnCoordSender](#TxnCoordSender) |
+| txn committed by coordinator | replay commit with writes in batch    | protected by WriteToOld errors                               |
+| txn committed by coordinator | replay commit without writes in batch | not an issue                                                 |
+| txn aborted by other txn     | replay commit                         | hit aborted txn record or TxnSpanGCThreshold                 |
```

Comment:
"replay commit" but also simply a race between the coordinator commit and a 3rd party abort, right?